src/compiler/scala/tools/nsc/PhaseAssembly.scala

```diff
@@ -12,271 +12,228 @@
 
-  /** Given the name of a phase object, get the node for that name.
-   *  If the node object does not exist, then create it.
-   */
-  def getNodeByPhase(name: String): Node      = nodes.getOrElseUpdate(name, Node(name))
-  def getNodeByPhase(phs: SubComponent): Node = {
-    val node: Node = getNodeByPhase(phs.phaseName)
-    if (node.phaseobj.isEmpty)
-      node.phaseobj = Some(List(phs))
-    node
-  }
+  /** For ith vertex, the count of its incoming edges. */
+  //private val inDegree: Array[Int] = Array.ofDim(order)
 
-  def softConnectNodes(frm: Node, to: Node) = connectNodes(Edge(frm, to, hard = false))
-  def hardConnectNodes(frm: Node, to: Node) = connectNodes(Edge(frm, to, hard = true))
+  /** Directed edge. */
+  private case class Edge(from: Int, to: Int, weight: Weight)
 
-  // Connect the frm and to nodes in the edge and add it to the set of edges.
-  private def connectNodes(e: Edge): Unit = {
-    edges += e
-    e.frm.after += e
-    e.to.before += e
-  }
+  // phase names and their vertex index
+  private val nodeCount = new java.util.concurrent.atomic.AtomicInteger
+  private val nodes = mutable.HashMap.empty[String, Int]
+  private val names = Array.ofDim[String](order)
 
-  /* Given the entire graph, collect the phase objects at each level, where the phase
-   * names are sorted alphabetical at each level, into the compiler phase list
+  /** Add the edge between named phases, where `to` follows `from`.
    */
-  def compilerPhaseList(): List[SubComponent] =
-    nodes.values.toList.filter(_.level > 0).sortBy(x => (x.level, x.phasename)).flatMap(_.phaseobj).flatten
-
-  // Test for graph cycles, assign levels to the nodes & collapse hard links into nodes.
-  def collapseHardLinksAndLevels(node: Node, lvl: Int)(implicit messaging: Messaging): Unit = {
-    if (node.visited) {
-      dump("phase-cycle")
-      throw new FatalError(s"Cycle in phase dependencies detected at ${node.phasename}, created phase-cycle.dot")
+  private def addEdge(from: String, to: String, weight: Weight): Unit = {
+    def getNode(name: String): Int = {
+      def installName(name: String, n: Int): Unit =
+        if (n >= names.length) throw new FatalError(names.mkString(s"Extra name $name; names [",",","]"))
+        else names(n) = name
+      nodes.getOrElseUpdate(name, nodeCount.getAndIncrement().tap(installName(name, _)))
     }
-
-    if (node.level < lvl)
-      node.level = lvl // level up
-    else if (node.level != 0) {
-      node.visited = false
-      return // no need to revisit
+    val v = getNode(from)
+    val w = getNode(to)
+    adjacency(v).find(e => e.from == v && e.to == w) match {
+      case None =>
+        //inDegree(w) += 1
+        //size += 1
+        adjacency(v) ::= Edge(from = v, to = w, weight)
+      case Some(e) if weight == FollowsNow => // use runsRightAfter edge FollowsNow
+        adjacency(v) = Edge(from = v, to = w, weight) :: adjacency(v).filterNot(e => e.from == v && e.to == w)
```

Comment:
There is already `v -> w` (runsBefore) and we add `v => w` (runsRightAfter), retaining the `=>` (for `FollowsNow`).

Otherwise, the new edge is normal `Follows` so keep existing edge, which can't have lower weight.